File: 2020-09-18/gioco_parentesi/sol/no_dfs.py
#!/usr/bin/python3
#Senza DFS. Notevole incremento prestazioni rispetto a normale traduzione.
import fileinput
import sys 
from collections import deque
def BUCKET(ID = -1, start = -1, end = -1, father = -1):
    a = [ID, start, end, father]
    return a

CACHE = []
ALL_BUCKETS = []

def init_globals(N):
    toReserve = int(N/2 + 1)
    global ALL_BUCKETS
    ALL_BUCKETS = [BUCKET() for _ in range(toReserve)]
    ALL_BUCKETS[0] = BUCKET(0, 0, -1, 0)
    global CACHE
    CACHE = [0 for _ in range(toReserve)]

def parser(parentesi):
    init_globals(len(parentesi))
    stack = deque()
    globalID = 0
    index = 0
    stack.appendleft(globalID)
    global ALL_BUCKETS
    global CACHE
    for c in parentesi:
        if(c == '('):
            globalID+=1
            ALL_BUCKETS[globalID] = BUCKET(globalID, index, -1, stack[0])
            stack.appendleft(globalID)
        elif(c == ')'):
            tempIndex = stack[0]
            ALL_BUCKETS[tempIndex][2] = index #end
            stack.popleft() #in pratica risalgo al padre
            ALL_BUCKETS[stack[0]][2] = index #aggiorno end del padre
            CACHE[stack[0]] ^= (CACHE[tempIndex] + 1) #aggiorno cache del padre
        else:
            exit(1)
        
        index += 1

if __name__ == "__main__":
    N = int(sys.stdin.readline())
    parentesi = sys.stdin.readline()
    parentesi = parentesi[:-1]
    assert(N == len(parentesi))

    parser(parentesi)

    risp = CACHE[0]

    if(0): #debug
        print()
        for _ in ALL_BUCKETS:
            print(_)
        print()
        print(CACHE)

    if(risp != 0):
        print(1)
        for i in range(1,len(CACHE)):
            j = i
            tryOne = CACHE[ALL_BUCKETS[j][3]] ^ (CACHE[j] + 1) #nulla i
            j = ALL_BUCKETS[j][3] #father
            while(j != 0): #ricalcola i valori fino alla radice
                tryOne = (tryOne + 1) ^ CACHE[ALL_BUCKETS[j][3]] ^ (CACHE[j] + 1)
                j = ALL_BUCKETS[j][3]
            if(tryOne == 0):
                print(ALL_BUCKETS[i][1], ALL_BUCKETS[i][2])
                break
    else:
        print(0)

File: 2020-09-18/gioco_parentesi/sol/no_dfs.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
#include <queue>

using BUCKET = struct {
                        size_t ID;
                        size_t start;
                        size_t end;
                        size_t father;
                      };

// PRETTY PRINT RUBATO - PER DEBUG PARSER
#include <algorithm>
#include <iterator>
#ifndef EVAL
std::ostream & operator<<(std::ostream & os, BUCKET bucket)
{
    os<<"[ ";
    os<< bucket.ID << "<" << bucket.father << ">:(" << bucket.start << "," << bucket.end << ")";
    os<<" ]";
    return os;
}

std::ostream & operator<<(std::ostream & os, BUCKET* bucket_ptr)
{
    if(bucket_ptr) {
        os<< *bucket_ptr;
    }
    return os;
}

template<typename T>
std::ostream & operator<<(std::ostream & os, std::vector<T> vec)
{
    os<<"{ ";
    for(auto it = vec.begin(); it != vec.end(); it++) os << *it << " ";
    os<<"}";
    return os;
}
#endif
//

using namespace std;

vector<BUCKET> ALL_BUCKETS;
vector<size_t> CACHE;

void init(size_t len) {
    size_t toReserve = len/2 + 1;
    CACHE = vector<size_t>(toReserve, 0);
    ALL_BUCKETS = vector<BUCKET>(toReserve);
}

void parser(string& s) {
    size_t globalID = 0;
    deque<size_t> stack;
    stack.push_front(0);
    init(s.size());
    size_t index = 0;
    for(auto c: s){
        if(c == '(') {
            globalID += 1;
            ALL_BUCKETS[globalID] = BUCKET({globalID,index,~0u,stack.front()});
            assert(ALL_BUCKETS[globalID].ID == globalID);
            stack.push_front(globalID);
        } else if (c == ')') {
            size_t tempIndex = stack.front();
            ALL_BUCKETS[tempIndex].end = index;
            stack.pop_front();
            ALL_BUCKETS[stack.front()].end = index;
            CACHE[stack.front()] ^= (CACHE[tempIndex] + 1);
        } else {
            exit(1);
        }
        index++;
    }
}

int main() {
    size_t len;
    cin >> len;
    string parentesi;
    cin >> parentesi;
    assert(len == parentesi.size());

    parser(parentesi);
    auto risp = CACHE[0];
#ifndef EVAL
    cout << ALL_BUCKETS << endl;
    cout << "RISP: " << risp  << endl << "DFS:" << endl;
    for(size_t i = 0; i < CACHE.size(); i++) {
        cout << "I: " << i << " --> " << CACHE[i] << endl;
    }
#endif
    cout << (bool)risp << endl;
    if(risp) {
        for(size_t i = 1; i < ALL_BUCKETS.size(); i++) {
            size_t j = i;
            size_t tryOne = CACHE[ALL_BUCKETS[j].father] ^ (CACHE[j] + 1);
            j = ALL_BUCKETS[j].father;
            while(j) {
                tryOne = (tryOne + 1) ^ CACHE[ALL_BUCKETS[j].father] ^ (CACHE[j] + 1);
                j = ALL_BUCKETS[j].father;
            }

            if(!tryOne) {
                cout << ALL_BUCKETS[i].start << " " << ALL_BUCKETS[i].end << endl;
                return 0;
            }
        }
    }

    return 0;
}